doc/man/man1/decrypt.1
.\" Automatically generated by Pandoc 1.16.0.2
.\"
.TH "DECRYPT" "1" "February 29, 2016" "Virgil Security CLI (2.0.0)" "Virgil"
.hy
.SH NAME
.PP
\f[B]decrypt\f[] \[en] decrypt data with given password or given Private
Key + recipient\-id
.SH SYNOPSIS
.IP
.nf
\f[C]
\ \ \ \ virgil\ decrypt\ \ [\-i\ <file>]\ [\-o\ <file>]\ [\-c\ <file>]\ [\-k\ <file>]\ [\-p\ <arg>]\ \-r
\ \ \ \ \ \ \ \ \ \ \ \ <arg>\ [\-V]\ [\-\-]\ [\-\-version]\ [\-h]
\f[]
.fi
.SH DESCRIPTION
.PP
Decrypt data with given password or given Private Key + recipient\-id.
recipient\-id is an identifier which is connected with Public Key.
If a \f[I]sender\f[] has a Card, his recipient\-id is Card's id.
.SH OPTIONS
.IP
.nf
\f[C]
\ \ \ \ \-i\ <file>,\ \ \-\-in\ <file>
\ \ \ \ \ Data\ to\ be\ decrypted.\ If\ omitted,\ stdin\ is\ used.

\ \ \ \ \-o\ <file>,\ \ \-\-out\ <file>
\ \ \ \ \ Decrypted\ data.\ If\ omitted,\ stdout\ is\ used.

\ \ \ \ \-c\ <file>,\ \ \-\-content\-info\ <file>
\ \ \ \ \ Content\ info.\ Use\ this\ option\ if\ content\ info\ is\ not\ embedded\ in\ the
\ \ \ \ \ encrypted\ data.

\ \ \ \ \-k\ <file>,\ \ \-\-key\ <file>
\ \ \ \ \ Private\ Key.

\ \ \ \ \-p\ <arg>,\ \ \-\-private\-key\-password\ <arg>
\ \ \ \ \ Password\ to\ be\ used\ for\ Private\ Key\ encryption.

\ \ \ \ \-r\ <arg>,\ \ \-\-recipient\ <arg>
\ \ \ \ \ (required)\ \ Recipient\ defined\ in\ format:

\ \ \ \ \ [password|id|vcard|email|private]:<value>

\ \ \ \ \ where:

\ \ \ \ \ \ \ \ *\ if\ password,\ then\ <value>\ \-\ recipient\[aq]s\ password;

\ \ \ \ \ \ \ \ *\ if\ id,\ then\ <value>\ \-\ recipient\[aq]s\ UUID\ associated\ with\ Virgil\ Card
\ \ \ \ \ identifier;

\ \ \ \ \ \ \ \ *\ if\ vcard,\ then\ <value>\ \-\ recipient\[aq]s\ Virgil\ Card/Cards\ file

\ \ \ \ \ \ \ \ \ \ stored\ locally;

\ \ \ \ \ \ \ \ *\ if\ email,\ then\ <value>\ \-\ recipient\[aq]s\ email;

\ \ \ \ \ \ \ \ *\ if\ private,\ then\ set\ type:value\ for\ searching\ Private\ Virgil\ Card[s].

\ \ \ \ \ For\ example:\ private:email:<obfuscator_email>.\ (\ obfiscator\ \-\ see
\ \ \ \ \ \[aq]virgil\ hash\[aq])


\ \ \ \ \-V,\ \ \-\-VERBOSE
\ \ \ \ \ Show\ detailed\ information

\ \ \ \ \-\-,\ \ \-\-ignore_rest
\ \ \ \ \ Ignores\ the\ rest\ of\ the\ labeled\ arguments\ following\ this\ flag.

\ \ \ \ \-\-version
\ \ \ \ \ Displays\ version\ information\ and\ exits.

\ \ \ \ \-h,\ \ \-\-help
\ \ \ \ \ Displays\ usage\ information\ and\ exits.
\f[]
.fi
.SH EXAMPLES
.IP "1." 3
Decrypt data for user identified by password:
.RS 4
.IP
.nf
\f[C]
virgil\ decrypt\ \-i\ plain.txt.enc\ \-o\ plain.txt\ \-r\ password:strong_password
\f[]
.fi
.RE
.IP "2." 3
Decrypt data for Bob identified by his Private Key + recipient\-id
[id|vcard|email|private]:
.RS 4
.IP
.nf
\f[C]
virgil\ decrypt\ \-i\ plain.txt.enc\ \-o\ plain.txt\ \-k\ bob/private.key\ \-r\ id:<recipient_id>
\f[]
.fi
.RE
.SH SEE ALSO
.PP
virgil(1)
.PD 0
.P
.PD
encrypt(1)
.SH AUTHORS
(c) Virgil Security Inc.


ChangeLog
virgil CLI ChangeLog (Sorted per date)

= Version v2.0.0-rc1 released 2016-05-16

Changes:
  * Use virgil-crypto lib 1.6.0
  * Use virgil-sdk-cpp lib 3.2.0
  * Add config file


= Version 0.3.4 released 2015-08-27

Bugfix:
  * Update underlying virgil crypto library to fixed version 1.0.1.

= Version 0.3.3 released 2015-08-18

Bugfix:
  * Embed MSVC redistributable package installation to the CLI package.

Changes:
  * Add executable folder to the PATH environment variable silently on Windows.

= Version 0.3.2 released 2015-08-17

Bugfix:
  * Fix RSA and EC keys generation.

= Version 0.3.1 released 2015-08-17

Bugfix:
  * Add missing win32 runtime libraries.

= Version 0.3.0 released 2015-08-13

Improvements:
  * Add Markdown documentation.

Changes:
  * Replace certificate term with Virgil Public Key term.
  * Remove options to specify sign's output format.
  * Use C++11 standard.

Features:
  * Add command "keyreg" - register user's public key on the Virgil Public Keys service.
  * Add command "keyget" - get user's public key from the Virgil Public Keys service.
  * Add command "confirm" - send confirmation code to the Virgil Public Keys service.
  * Add command "reconfirm" - resend confirmation code to the user for given user's identifier.

= Version 0.2.0 released 2015-05-15

Bugfix:
  * Use standard constructor of class 'ifstream'.
  * Change parameter 'p' long name from 'arg' to 'pwd'.
  * Add required include statement <cstdlib>.

Improvements:
  * Fix misspelling in the help text.
  * Add "happy path" test.
  * Modify 'encrypt' and 'verify' commands to work with Virgil PKI service thru user's identifier.

Changes:
  * Change parameter 'asn1' to 'der'.
  * Change parameters format for 'encrypt' command.

Features:
  * Add 'verify' command.
  * Add 'adduser' command

= Version 0.1.0 released 2015-05-04

Features:
  * Add command "keygen" - generate private key with given parameters.
  * Add command "key2pub" - extract public key from the private key.
  * Add command "pub2cert" - create certificate from the public key and identifiers.
  * Add command "certinfo" - output detailed information about given certificate.
  * Add command "encrypt" - encrypt data for given recipients which can be defined by certificates and by passwords.
  * Add command "decrypt" - decrypt data for given recipient which can be defined by certificate or by password.
  * Add command "sign" - sign data with private key.
  * Add command "verify" - verify data with certificate.


src/commands/base/decrypt.cxx
/**
 * Copyright (C) 2015 Virgil Security Inc.
 *
 * Lead Maintainer: Virgil Security Inc. <support@virgilsecurity.com>
 *
 * All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions are
 * met:
 *
 *     (1) Redistributions of source code must retain the above copyright
 *     notice, this list of conditions and the following disclaimer.
 *
 *     (2) Redistributions in binary form must reproduce the above copyright
 *     notice, this list of conditions and the following disclaimer in
 *     the documentation and/or other materials provided with the
 *     distribution.
 *
 *     (3) Neither the name of the copyright holder nor the names of its
 *     contributors may be used to endorse or promote products derived from
 *     this software without specific prior written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY THE AUTHOR ''AS IS'' AND ANY EXPRESS OR
 * IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
 * WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
 * DISCLAIMED. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY DIRECT,
 * INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
 * (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
 * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
 * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT,
 * STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING
 * IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
 * POSSIBILITY OF SUCH DAMAGE.
 */

#include <fstream>
#include <iostream>
#include <stdexcept>
#include <string>
#include <vector>
#include <iterator>

#include <tclap/CmdLine.h>

#include <virgil/crypto/VirgilByteArray.h>

#include <virgil/crypto/VirgilCipher.h>
#include <virgil/crypto/VirgilStreamCipher.h>
#include <virgil/crypto/stream/VirgilStreamDataSource.h>
#include <virgil/crypto/stream/VirgilStreamDataSink.h>

#include <cli/pair.h>
#include <cli/version.h>
#include <cli/util.h>

namespace vcrypto = virgil::crypto;
namespace vsdk = virgil::sdk;
namespace vcli = virgil::cli;

#ifdef SPLIT_CLI
#define MAIN main
#else
#define MAIN decrypt_main
#endif

static void reset(std::istream& in);

int MAIN(int argc, char** argv) {
    try {
        std::string description = "Decrypt data with given password or given Private Key + recipient-id. "
                                  "recipient-id is an identifier which is connected with Public Key. "
                                  "If a sender has a Card, his recipient-id is Card's id. Also, Public "
                                  "Key is saved in the Card.\n\n";

        std::vector<std::string> examples;
        examples.push_back("Decrypt data for user identified by password:\n"
                           "virgil decrypt -i plain.txt.enc -o plain.txt -r password:strong_password\n\n");

        examples.push_back(
            "Decrypt data for Bob identified by his Private Key + recipient-id [id|vcard|email|private]:\n"
            "virgil decrypt -i plain.txt.enc -o plain.txt -k bob/private.key -r id:<recipient_id>\n\n");

        std::string descriptionMessage = virgil::cli::getDescriptionMessage(description, examples);

        // Parse arguments.
        TCLAP::CmdLine cmd(descriptionMessage, ' ', virgil::cli_version());

        TCLAP::ValueArg<std::string> inArg("i", "in", "Data to be decrypted. If omitted, stdin is used.", false, "",
                                           "file");

        TCLAP::ValueArg<std::string> outArg("o", "out", "Decrypted data. If omitted, stdout is used.", false, "",
                                            "file");

        TCLAP::ValueArg<std::string> contentInfoArg("c", "content-info",
                                                    "Content info. Use this option if"
                                                    " content info is not embedded in the encrypted data.",
                                                    false, "", "file");

        TCLAP::ValueArg<std::string> privateKeyArg("k", "key", "Private Key.", false, "", "file");

        TCLAP::ValueArg<std::string> privateKeyPasswordArg(
            "p", "private-key-password", "Password to be used for Private Key encryption.", false, "", "arg");

        TCLAP::ValueArg<std::string> recipientArg(
            "r", "recipient", "Recipient defined in format:\n"
                              "[password|id|vcard|email|private]:<value>\n"
                              "where:\n"
                              "\t* if password, then <value> - recipient's password;\n"
                              "\t* if id, then <value> - recipient's UUID associated with Virgil Card identifier;\n"
                              "\t* if vcard, then <value> - recipient's Virgil Card/Cards file\n\t  stored locally;\n"
                              "\t* if email, then <value> - recipient's email;\n"
                              "\t* if private, then set type:value for searching Private Virgil Card[s].\n"
                              "For example:\n"
                              "1. private:<obfuscator_type>:<obfuscator_email>. ( obfiscator - see 'virgil hash')\n"
                              "2. private:<identity_type>:<identity_value>; private:email:alice@domain.com\n",
            true, "", "arg");

        TCLAP::SwitchArg verboseArg("V", "VERBOSE", "Show detailed information", false);

        cmd.add(verboseArg);
        cmd.add(recipientArg);
        cmd.add(privateKeyPasswordArg);
        cmd.add(privateKeyArg);
        cmd.add(contentInfoArg);
        cmd.add(outArg);
        cmd.add(inArg);
        cmd.parse(argc, argv);

        auto recipientFormat = vcli::parsePair(recipientArg.getValue());
        vcli::checkFormatRecipientArg(recipientFormat);

        // Create cipher
        vcrypto::VirgilStreamCipher cipher;

        if (!contentInfoArg.getValue().empty()) {
            // Set content info.
            std::ifstream contentInfoFile(contentInfoArg.getValue(), std::ios::in | std::ios::binary);
            if (!contentInfoFile) {
                throw std::invalid_argument("cannot read file: " + contentInfoArg.getValue());
            }

            vcrypto::VirgilByteArray contentInfo((std::istreambuf_iterator<char>(contentInfoFile)),
                                                 std::istreambuf_iterator<char>());
            cipher.setContentInfo(contentInfo);
        }

        // Prepare input
        std::istream* inStream;
        std::ifstream inFile;
        if (inArg.getValue().empty() || inArg.getValue() == "-") {
            inStream = &std::cin;
        } else {
            inFile.open(inArg.getValue(), std::ios::in | std::ios::binary);
            if (!inFile) {
                throw std::invalid_argument("cannot read file: " + inArg.getValue());
            }
            inStream = &inFile;
        }

        // Prepare output
        std::ostream* outStream;
        std::ofstream outFile;
        if (outArg.getValue().empty()) {
            outStream = &std::cout;
        } else {
            outFile.open(outArg.getValue(), std::ios::out | std::ios::binary);
            if (!outFile) {
                throw std::invalid_argument("cannot write file: " + outArg.getValue());
            }
            outStream = &outFile;
        }

        // Create IO streams
        vcrypto::stream::VirgilStreamDataSource dataSource(*inStream);
        vcrypto::stream::VirgilStreamDataSink dataSink(*outStream);

        std::string type = recipientFormat.first;
        std::string value = recipientFormat.second;

        if (type == "password") {
            vcrypto::VirgilByteArray password = virgil::crypto::str2bytes(value);
            cipher.decryptWithPassword(dataSource, dataSink, password);
            if (verboseArg.isSet()) {
                std::cout << "File has been decrypted with a password" << std::endl;
            }
        } else {
            // type = [id|vcard|email]
            // Read private key
            std::string pathToPrivateKeyFile = privateKeyArg.getValue();
            vcrypto::VirgilByteArray privateKey = vcli::readPrivateKey(pathToPrivateKeyFile);

            vcrypto::VirgilByteArray privateKeyPassword;
            if (privateKeyPasswordArg.isSet()) {
                privateKeyPassword = vcrypto::str2bytes(privateKeyPasswordArg.getValue());
            } else {
                privateKeyPassword = vcli::setPrivateKeyPass(privateKey);
            }

            // type = [id|vcard|email|private]
            // if recipient email:<value>, then a download Virgil Card with confirmed identity
            if (type == "id") {
                std::string recipientCardId = value;
                cipher.decryptWithKey(dataSource, dataSink, vcrypto::str2bytes(recipientCardId), privateKey,
                                      privateKeyPassword);
                return EXIT_SUCCESS;
            }

            std::vector<std::string> recipientCardsId;
            if (type == "private") {
                // private:<type>:<value>
                auto pairTypeAndValue = vcli::parsePair(value);
                std::string type = pairTypeAndValue.first;
                std::string value = pairTypeAndValue.second;

                bool isSearchPrivateCard = true; // search the Private Virgil Card[s]
                recipientCardsId = vcli::getRecipientCardsId(verboseArg.isSet(), type, value, isSearchPrivateCard);
            } else {
                // type = [id|vcard|email]
                bool isSearchPrivateCard = false; // search the Global Virgil Card[s]
                recipientCardsId = vcli::getRecipientCardsId(verboseArg.isSet(), type, value, isSearchPrivateCard);
            }

            if (recipientCardsId.empty()) {
                if (verboseArg.isSet()) {
                    std::cout << "Cards by " << type << ":" << value << " haven't been found." << std::endl;
                    return EXIT_FAILURE;
                }
            }

            size_t countErrorDecryptWithKey = 0;
            vcrypto::VirgilByteArray encryptedData((std::istreambuf_iterator<char>(inStream->rdbuf())),
                                                   std::istreambuf_iterator<char>());

            for (const auto& recipientCardId : recipientCardsId) {
                try {
                    if (inArg.getValue().empty() || inArg.getValue() == "-") {
                        vcrypto::VirgilCipher cipher;
                        vcrypto::VirgilByteArray decryptedData = cipher.decryptWithKey(
                            encryptedData, vcrypto::str2bytes(recipientCardId), privateKey, privateKeyPassword);

                        vcli::writeBytes(outArg.getValue(), decryptedData);
                        break;
                    } else {
                        reset(*inStream);
                        vcrypto::VirgilStreamCipher streamCipher;
                        streamCipher.decryptWithKey(dataSource, dataSink, vcrypto::str2bytes(recipientCardId),
                                                    privateKey, privateKeyPassword);
                        break;
                    }

                } catch (std::exception& exception) {
                    ++countErrorDecryptWithKey;
                    if (verboseArg.isSet()) {
                        std::cout << "decrypt. Warning: " << exception.what() << std::endl;
                    }
                }
            }

            if (countErrorDecryptWithKey == recipientCardsId.size()) {
                throw std::runtime_error("File can’t be decrypted:\n"
                                         "wrong card-id or/and private key.\n");
            }

            if (verboseArg.isSet()) {
                std::cout << "File has been decrypted with a private key" << std::endl;
            }
        }

    } catch (TCLAP::ArgException& exception) {
        std::cerr << "decrypt. Error: " << exception.error() << " for arg " << exception.argId() << std::endl;
        return EXIT_FAILURE;
    } catch (std::exception& exception) {
        std::cerr << "decrypt. Error: " << exception.what() << std::endl;
        return EXIT_FAILURE;
    }

    return EXIT_SUCCESS;
}

void reset(std::istream& in) {
    in.clear();
    in.seekg(0, in.beg);
}
